File: solutions/objective-c/bob/1/Bob.m
#import <Foundation/Foundation.h>
#import "Bob.h"
@implementation Bob
-(BOOL) hasLetter:(NSString *)string {
    return [string rangeOfCharacterFromSet: [NSCharacterSet letterCharacterSet]].location != NSNotFound;
}
- (NSString *)hey:(NSString *)input {
    BOOL isEmptyString = [[input stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]] length] == 0;
    NSCharacterSet *alphabetSet = [NSCharacterSet letterCharacterSet];
    BOOL isAllUppercase = [input.uppercaseString isEqualToString:input] && [self hasLetter:input];
    if(isEmptyString) {
        return @"Fine. Be that way!";
    } else if(isAllUppercase) {
        return @"Whoa, chill out!";
    } else if( [input hasSuffix:@"?"]) {
        return @"Sure.";
    } else {
        return @"Whatever.";
    }
}
@end

File: solutions/delphi/bob/1/bob.pas
unit uBob;

interface

type
  TBob = class
  public
    class function Hey(Value: string): string;
  end;

implementation

uses
  System.RegularExpressions,
  System.StrUtils,
  System.SysUtils;

class function TBob.Hey(Value: string): string;
begin
  Value := Value.Trim;
  if TRegEx.IsMatch(Value, '[A-Z]') and not TRegEx.IsMatch(Value, '[a-z]') then
    Result := 'Whoa, chill out!'
  else if Value.EndsWith('?') then
    Result := 'Sure.'
  else if Value = '' then
    Result := 'Fine.  Be that way!'   
  else
    Result := 'Whatever.';
end;

end.